Feature request: `http_only` compression middleware

Currently using jsonrpsee http_only server and wanted to compress the response before returning to the client using tower-http CompressionLayer.

referencing: [CorsLayer](https://github.com/paritytech/jsonrpsee/blob/master/examples/examples/cors_server.rs)

```rust
use tower_http::compression::CompressionLayer;

let service_builder = ServiceBuilder::new()
        .layer(ProxyGetRequestLayer::new("/health", "system_health")?)
        .layer(CompressionLayer::new());

let server = ServerBuilder::default()
        .http_only()
        .set_host_filtering(AllowHosts::Any)
        .set_middleware(service_builder)
        .build("127.0.0.1:3000".parse::<SocketAddr>()?)
        .await?;

let mut module = RpcModule::new(());
/* register methods code block */

let handle = server.start(module)?;
```

After adding CompressionLayer it throw the following errors
```rust
55 |     let handle = server.start(module)?;
   |                         ^^^^^ method cannot be called due to unsatisfied trait bounds
   |
   |
62 | pub struct ProxyGetRequest<S> {
   | -----------------------------
   | |
   | doesn't satisfy `<_ as Service<Request<Body>>>::Error = Box<(dyn Error + Send + Sync + 'static)>`
   | doesn't satisfy `<_ as Service<Request<Body>>>::Response = Response<_>`
   | doesn't satisfy `_: Service<Request<Body>>`
   |
16 | pub struct Compression<S, P = DefaultPredicate> {
   | ----------------------------------------------- doesn't satisfy `<_ as Service<Request<Body>>>::Response = Response<Body>`
   |
   = note: the following trait bounds were not satisfied:
           `<ProxyGetRequest<Compression<jsonrpsee::jsonrpsee_server::server::TowerService<()>>> as Service<http::request::Request<jsonrpsee::jsonrpsee_server::logger::Body>>>::Response = http::response::Response<_>`
           `<ProxyGetRequest<Compression<jsonrpsee::jsonrpsee_server::server::TowerService<()>>> as Service<http::request::Request<jsonrpsee::jsonrpsee_server::logger::Body>>>::Error = Box<(dyn StdError + std::marker::Send + std::marker::Sync + 'static)>`
           `ProxyGetRequest<Compression<jsonrpsee::jsonrpsee_server::server::TowerService<()>>>: Service<http::request::Request<jsonrpsee::jsonrpsee_server::logger::Body>>`
           `<Compression<jsonrpsee::jsonrpsee_server::server::TowerService<()>> as Service<http::request::Request<jsonrpsee::jsonrpsee_server::logger::Body>>>::Response = http::response::Response<jsonrpsee::jsonrpsee_server::logger::Body>`
```

If placed the layers differently i'll get a different error

```rust
// CompressionLayer first before ProxyGetRequestLayer

let service_builder = ServiceBuilder::new()
        .layer(CompressionLayer::new())
        .layer(ProxyGetRequestLayer::new("/health", "system_health")?);
```

Build error
```rust
error[E0277]: the size for values of type `(dyn StdError + std::marker::Send + std::marker::Sync + 'static)` cannot be known at compilation time
   |
56 |     let handle = server.start(module)?;
   |                         ^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `(dyn StdError + std::marker::Send + std::marker::Sync + 'static)`
   = help: the trait `StdError` is implemented for `Box<T>`
   = note: required for `Box<(dyn StdError + std::marker::Send + std::marker::Sync + 'static)>` to implement `StdError`
note: required by a bound in `jsonrpsee::jsonrpsee_server::Server::<S, L>::start`
   |
99 |     <B as HttpBody>::Error: Send + Sync + StdError,
   |                                           ^^^^^^^^ required by this bound in `Server::<S, L>::start`
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@ The format is based on [Keep a Changelog].
 
 ## [v0.24.7]
 
+This is a bug-fix release that fixes the tower::Service implementation to be generic over the HttpBody to work with all middleware layers. For instance, this makes `tower_http::compression::CompressionLayer` work, which didn't compile before.
 ### [Fixed]
 - fix(server): make tower::Service impl generic over HttpBody ([#1475](https://github.com/paritytech/jsonrpsee/pull/1475))
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,7 +7,7 @@ The format is based on [Keep a Changelog].
 ## [v0.24.7]
 
 ### [Fixed]
-- fix(server): compatibility issue of CompressionLayer in http middleware ([#1475](https://github.com/paritytech/jsonrpsee/pull/1475))
+- fix(server): make tower::Service impl generic over HttpBody ([#1475](https://github.com/paritytech/jsonrpsee/pull/1475))
 
 ## [v0.24.6] - 2024-10-07
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,11 @@ The format is based on [Keep a Changelog].
 
 [Keep a Changelog]: http://keepachangelog.com/en/1.0.0/
 
+## Unrelease
+
+### [Fixed]
+- fix(server): compatibility issue of CompressionLayer in http middleware ([#1475](https://github.com/paritytech/jsonrpsee/pull/1475))
+
 ## [v0.24.6] - 2024-10-07
 
 This is a bug-fix release that fixes that the `ConnectionGuard` was dropped before the future was resolved which,

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,7 +6,8 @@ The format is based on [Keep a Changelog].
 
 ## [v0.24.7]
 
-This is a bug-fix release that fixes the tower::Service implementation to be generic over the HttpBody to work with all middleware layers. For instance, this makes `tower_http::compression::CompressionLayer` work, which didn't compile before.
+This is a bug-fix release that fixes the tower::Service implementation to be generic over the HttpBody to work with all middleware layers. 
+For instance, this makes `tower_http::compression::CompressionLayer` work, which didn't compile before.
 ### [Fixed]
 - fix(server): make tower::Service impl generic over HttpBody ([#1475](https://github.com/paritytech/jsonrpsee/pull/1475))
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,7 @@ The format is based on [Keep a Changelog].
 
 This is a bug-fix release that fixes the tower::Service implementation to be generic over the HttpBody to work with all middleware layers. 
 For instance, this makes `tower_http::compression::CompressionLayer` work, which didn't compile before.
+
 ### [Fixed]
 - fix(server): make tower::Service impl generic over HttpBody ([#1475](https://github.com/paritytech/jsonrpsee/pull/1475))
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,7 +4,7 @@ The format is based on [Keep a Changelog].
 
 [Keep a Changelog]: http://keepachangelog.com/en/1.0.0/
 
-## Unrelease
+## [v0.24.7]
 
 ### [Fixed]
 - fix(server): compatibility issue of CompressionLayer in http middleware ([#1475](https://github.com/paritytech/jsonrpsee/pull/1475))

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -19,7 +19,10 @@ members = [
 resolver = "2"
 
 [workspace.package]
-authors = ["Parity Technologies <admin@parity.io>", "Pierre Krieger <pierre.krieger1708@gmail.com>"]
+authors = [
+	"Parity Technologies <admin@parity.io>",
+	"Pierre Krieger <pierre.krieger1708@gmail.com>",
+]
 version = "0.24.6"
 edition = "2021"
 rust-version = "1.74.1"
@@ -40,5 +43,5 @@ jsonrpsee-wasm-client = { path = "client/wasm-client", version = "0.24.6" }
 jsonrpsee-client-transport = { path = "client/transport", version = "0.24.6" }
 jsonrpsee-proc-macros = { path = "proc-macros", version = "0.24.6" }
 
-tower = "0.5"
-tower-http = "0.6"
+tower = "0.4"
+tower-http = "0.5"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -19,11 +19,8 @@ members = [
 resolver = "2"
 
 [workspace.package]
-authors = [
-	"Parity Technologies <admin@parity.io>",
-	"Pierre Krieger <pierre.krieger1708@gmail.com>",
-]
-version = "0.24.6"
+authors = ["Parity Technologies <admin@parity.io>", "Pierre Krieger <pierre.krieger1708@gmail.com>"]
+version = "0.24.7"
 edition = "2021"
 rust-version = "1.74.1"
 license = "MIT"
@@ -34,14 +31,14 @@ keywords = ["jsonrpc", "json", "http", "websocket", "WASM"]
 readme = "README.md"
 
 [workspace.dependencies]
-jsonrpsee-types = { path = "types", version = "0.24.6" }
-jsonrpsee-core = { path = "core", version = "0.24.6" }
-jsonrpsee-server = { path = "server", version = "0.24.6" }
-jsonrpsee-ws-client = { path = "client/ws-client", version = "0.24.6" }
-jsonrpsee-http-client = { path = "client/http-client", version = "0.24.6" }
-jsonrpsee-wasm-client = { path = "client/wasm-client", version = "0.24.6" }
-jsonrpsee-client-transport = { path = "client/transport", version = "0.24.6" }
-jsonrpsee-proc-macros = { path = "proc-macros", version = "0.24.6" }
+jsonrpsee-types = { path = "types", version = "0.24.7" }
+jsonrpsee-core = { path = "core", version = "0.24.7" }
+jsonrpsee-server = { path = "server", version = "0.24.7" }
+jsonrpsee-ws-client = { path = "client/ws-client", version = "0.24.7" }
+jsonrpsee-http-client = { path = "client/http-client", version = "0.24.7" }
+jsonrpsee-wasm-client = { path = "client/wasm-client", version = "0.24.7" }
+jsonrpsee-client-transport = { path = "client/transport", version = "0.24.7" }
+jsonrpsee-proc-macros = { path = "proc-macros", version = "0.24.7" }
 
 tower = "0.4"
 tower-http = "0.5"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -39,3 +39,6 @@ jsonrpsee-http-client = { path = "client/http-client", version = "0.24.6" }
 jsonrpsee-wasm-client = { path = "client/wasm-client", version = "0.24.6" }
 jsonrpsee-client-transport = { path = "client/transport", version = "0.24.6" }
 jsonrpsee-proc-macros = { path = "proc-macros", version = "0.24.6" }
+
+tower = "0.5"
+tower-http = "0.6"

--- a/README.md
+++ b/README.md
@@ -6,7 +6,7 @@
 ![MIT](https://img.shields.io/crates/l/jsonrpsee.svg)
 [![CI](https://github.com/paritytech/jsonrpsee/actions/workflows/ci.yml/badge.svg)](https://github.com/paritytech/jsonrpsee/actions/workflows/ci.yml)
 [![Benchmarks](https://github.com/paritytech/jsonrpsee/actions/workflows/benchmarks_gitlab.yml/badge.svg)](https://github.com/paritytech/jsonrpsee/actions/workflows/benchmarks_gitlab.yml)
-[![dependency status](https://deps.rs/crate/jsonrpsee/0.24.6/status.svg)](https://deps.rs/crate/jsonrpsee/0.24.6)
+[![dependency status](https://deps.rs/crate/jsonrpsee/0.24.7/status.svg)](https://deps.rs/crate/jsonrpsee/0.24.7)
 
 JSON-RPC library designed for async/await in Rust.
 

--- a/README.md
+++ b/README.md
@@ -6,7 +6,7 @@
 ![MIT](https://img.shields.io/crates/l/jsonrpsee.svg)
 [![CI](https://github.com/paritytech/jsonrpsee/actions/workflows/ci.yml/badge.svg)](https://github.com/paritytech/jsonrpsee/actions/workflows/ci.yml)
 [![Benchmarks](https://github.com/paritytech/jsonrpsee/actions/workflows/benchmarks_gitlab.yml/badge.svg)](https://github.com/paritytech/jsonrpsee/actions/workflows/benchmarks_gitlab.yml)
-[![dependency status](https://deps.rs/crate/jsonrpsee/0.24.7/status.svg)](https://deps.rs/crate/jsonrpsee/0.24.7)
+[![dependency status](https://deps.rs/crate/jsonrpsee/latest/status.svg)](https://deps.rs/crate/jsonrpsee)
 
 JSON-RPC library designed for async/await in Rust.
 

--- a/client/http-client/Cargo.toml
+++ b/client/http-client/Cargo.toml
@@ -29,7 +29,7 @@ serde_json = "1"
 thiserror = "1"
 tokio = { version = "1.23.1", features = ["time"] }
 tracing = "0.1.34"
-tower = { version = "0.4.13", features = ["util"] }
+tower = { workspace = true, features = ["util"] }
 url = "2.4.0"
 
 [dev-dependencies]

--- a/examples/Cargo.toml
+++ b/examples/Cargo.toml
@@ -19,7 +19,7 @@ tokio = { version = "1.23.1", features = ["full"] }
 tokio-stream = { version = "0.1", features = ["sync"] }
 serde_json = { version = "1" }
 tower-http = { version = "0.5.2", features = ["full"] }
-tower = { version = "0.4.13", features = ["full"] }
+tower = { workspace = true, features = ["full"] }
 hyper = "1.3"
 hyper-util = { version = "0.1.3", features = ["client", "client-legacy"]}
 console-subscriber = "0.4.0"

--- a/examples/examples/http_middleware.rs
+++ b/examples/examples/http_middleware.rs
@@ -40,6 +40,7 @@ use jsonrpsee::rpc_params;
 use std::iter::once;
 use std::net::SocketAddr;
 use std::time::Duration;
+use tower_http::compression::CompressionLayer;
 use tower_http::cors::CorsLayer;
 use tower_http::sensitive_headers::SetSensitiveRequestHeadersLayer;
 use tower_http::trace::{DefaultMakeSpan, DefaultOnResponse, TraceLayer};
@@ -107,6 +108,7 @@ async fn run_server() -> anyhow::Result<SocketAddr> {
 		// Mark the `Authorization` request header as sensitive so it doesn't show in logs
 		.layer(SetSensitiveRequestHeadersLayer::new(once(hyper::header::AUTHORIZATION)))
 		.layer(cors)
+		.layer(CompressionLayer::new())
 		.timeout(Duration::from_secs(2));
 
 	let server =

--- a/examples/examples/jsonrpsee_as_service.rs
+++ b/examples/examples/jsonrpsee_as_service.rs
@@ -48,6 +48,7 @@ use jsonrpsee::ws_client::{HeaderValue, WsClientBuilder};
 use jsonrpsee::{MethodResponse, Methods};
 use tokio::net::TcpListener;
 use tower::Service;
+use tower_http::compression::CompressionLayer;
 use tower_http::cors::CorsLayer;
 use tracing_subscriber::util::SubscriberInitExt;
 
@@ -202,6 +203,7 @@ async fn run_server(metrics: Metrics) -> anyhow::Result<ServerHandle> {
 				let transport_label = if is_websocket { "ws" } else { "http" };
 				let PerConnection { methods, stop_handle, metrics, svc_builder } = per_conn2.clone();
 
+				let http_middleware = tower::ServiceBuilder::new().layer(CompressionLayer::new());
 				// NOTE, the rpc middleware must be initialized here to be able to created once per connection
 				// with data from the connection such as the headers in this example
 				let headers = req.headers().clone();
@@ -209,7 +211,10 @@ async fn run_server(metrics: Metrics) -> anyhow::Result<ServerHandle> {
 					AuthorizationMiddleware { inner: service, headers: headers.clone(), transport_label }
 				});
 
-				let mut svc = svc_builder.set_rpc_middleware(rpc_middleware).build(methods, stop_handle);
+				let mut svc = svc_builder
+					.set_http_middleware(http_middleware)
+					.set_rpc_middleware(rpc_middleware)
+					.build(methods, stop_handle);
 
 				if is_websocket {
 					// Utilize the session close future to know when the actual WebSocket

--- a/proc-macros/Cargo.toml
+++ b/proc-macros/Cargo.toml
@@ -33,4 +33,4 @@ serde_json = "1"
 serde = "1"
 trybuild = "1.0.97"
 tokio = { version = "1.23.1", features = ["rt", "macros"] }
-tower = "0.4"
+tower = { workspace = true }

--- a/server/Cargo.toml
+++ b/server/Cargo.toml
@@ -29,7 +29,7 @@ hyper-util = { version = "0.1", features = ["tokio", "service", "tokio", "server
 http = "1"
 http-body = "1"
 http-body-util = "0.1.0"
-tower = { version = "0.4.13", features = ["util"] }
+tower = { workspace = true, features = ["util"] }
 thiserror = "1"
 route-recognizer = "0.3.1"
 pin-project = "1.1.3"
@@ -37,5 +37,5 @@ pin-project = "1.1.3"
 [dev-dependencies]
 jsonrpsee-test-utils = { path = "../test-utils" }
 tracing-subscriber = { version = "0.3.3", features = ["env-filter"] }
-tower = { version = "0.4.13", features = ["timeout"] }
+tower = { workspace = true, features = ["timeout"] }
 socket2 = "0.5.1"

--- a/server/src/server.rs
+++ b/server/src/server.rs
@@ -970,20 +970,20 @@ impl<RpcMiddleware, HttpMiddleware> TowerService<RpcMiddleware, HttpMiddleware> 
 	}
 }
 
-impl<Body, RpcMiddleware, HttpMiddleware> Service<HttpRequest<Body>> for TowerService<RpcMiddleware, HttpMiddleware>
+impl<RequestBody, ResponseBody, RpcMiddleware, HttpMiddleware> Service<HttpRequest<RequestBody>> for TowerService<RpcMiddleware, HttpMiddleware>
 where
 	RpcMiddleware: for<'a> tower::Layer<RpcService> + Clone,
 	<RpcMiddleware as Layer<RpcService>>::Service: Send + Sync + 'static,
 	for<'a> <RpcMiddleware as Layer<RpcService>>::Service: RpcServiceT<'a>,
 	HttpMiddleware: Layer<TowerServiceNoHttp<RpcMiddleware>> + Send + 'static,
 	<HttpMiddleware as Layer<TowerServiceNoHttp<RpcMiddleware>>>::Service:
-		Send + Service<HttpRequest<Body>, Response = HttpResponse, Error = Box<(dyn StdError + Send + Sync + 'static)>>,
-	<<HttpMiddleware as Layer<TowerServiceNoHttp<RpcMiddleware>>>::Service as Service<HttpRequest<Body>>>::Future:
+		Send + Service<HttpRequest<RequestBody>, Response = HttpResponse<ResponseBody>, Error = Box<(dyn StdError + Send + Sync + 'static)>>,
+	<<HttpMiddleware as Layer<TowerServiceNoHttp<RpcMiddleware>>>::Service as Service<HttpRequest<RequestBody>>>::Future:
 		Send + 'static,
-	Body: http_body::Body<Data = Bytes> + Send + 'static,
-	Body::Error: Into<BoxError>,
+	RequestBody: http_body::Body<Data = Bytes> + Send + 'static,
+	RequestBody::Error: Into<BoxError>,
 {
-	type Response = HttpResponse;
+	type Response = HttpResponse<ResponseBody>;
 	type Error = BoxError;
 	type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
 
@@ -991,7 +991,7 @@ where
 		Poll::Ready(Ok(()))
 	}
 
-	fn call(&mut self, request: HttpRequest<Body>) -> Self::Future {
+	fn call(&mut self, request: HttpRequest<RequestBody>) -> Self::Future {
 		Box::pin(self.http_middleware.service(self.rpc_middleware.clone()).call(request))
 	}
 }

--- a/tests/Cargo.toml
+++ b/tests/Cargo.toml
@@ -22,8 +22,8 @@ serde_json = "1"
 tokio = { version = "1.23.1", features = ["full"] }
 tokio-stream = "0.1"
 tokio-util = { version = "0.7", features = ["compat"]}
-tower = { version = "0.4.13", features = ["full"] }
-tower-http = { version = "0.5.2", features = ["full"] }
+tower = { workspace = true, features = ["full"] }
+tower-http = { workspace = true, features = ["full"] }
 tracing = "0.1.34"
 tracing-subscriber = { version = "0.3.3", features = ["env-filter"] }
 pin-project = "1"

